src/main2.rs

```diff
@@ -0,0 +1,53 @@
+use cpal::{traits::{DeviceTrait, HostTrait, StreamTrait}, StreamConfig};
+use rustfft::{num_complex::Complex, FftPlanner};
+
+fn main() {
+    let stream = build_stream();
+    loop {
+    stream.play().expect("stream couldn't play?");
+    }
+}
+
+fn build_stream() -> cpal::Stream {
+    let host = cpal::default_host();
+    let my_dev = host.default_input_device().expect("no default input device found");
+    let config: StreamConfig = my_dev.default_input_config().expect("No default input config").into();
+    let sample_rate = config.sample_rate.0;
+    println!("Sample rate: {}", sample_rate);
+    let mut stream_acc: Vec<f32> = Vec::new();
+    let stream = my_dev.build_input_stream(&config,
+        move |data: &[f32], _: &cpal::InputCallbackInfo| {
+            stream_acc.extend(data.to_vec());
```

Comment:
You do not need to do `data.to_vec()` here. That is wasteful because it makes an extra copy of the data. You should be able to do `stream_acc.extend_from_slice(data)`. That will also eliminate the need for the `stream_acc: Vec<f32>` type annotation above

Reply:
Where do you learn about such nuances in built-ins and methods available in rust?

Reply:
Generally, that's something that comes with experience. I'm assuming that you've already seen it, but if you haven't, the Rust library docs are an amazing resource. For example, here's [the method on the Vec page for extend_from_slice](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.extend_from_slice).

To be honest Rust probably has the best documentation of any programming language that I've used. Other than that, I'm not really sure what to recommend haha